sign/api.go

```diff
@@ -0,0 +1,413 @@
+
+	// Creates a signature using the PrivateKey on the given message and
+	// returns the signature. opts are additional options which can be nil.
+	Sign(sk PrivateKey, message []byte, opts *SignatureOpts) []byte
+
+	// Checks whether the given signature is a valid signature set by
+	// the private key corresponding to the given public key on the
+	// given message. opts are additional options which can be nil.
+	Verify(pk PublicKey, message []byte, signature []byte, opts *SignatureOpts) bool
+
+	// Deterministically derives a keypair from a seed.  If you're unsure,
+	// you're better off using GenerateKey().
+	//
+	// Panics if seed is not of length SeedSize().
+	DeriveKey(seed []byte) (PublicKey, PrivateKey)
+
+	// Unmarshals a PublicKey from the provided buffer.
+	UnmarshalBinaryPublicKey([]byte) (PublicKey, error)
+
+	// Unmarshals a PublicKey from the provided buffer.
+	UnmarshalBinaryPrivateKey([]byte) (PrivateKey, error)
+
+	// Size of binary marshalled public keys
+	PublicKeySize() uint
+
+	// Size of binary marshalled public keys
+	PrivateKeySize() uint
+
+	// Name of the scheme
+	Name() string
+
+	// Size of signatures
+	SignatureSize() uint
+
+	// Size of seeds
+	SeedSize() uint
+}
+
+// Additional methods when the signature scheme is supported in X509.
+type CertificateScheme interface {
+	// Return the appropriate OIDs for this instance.  It is implicitly
+	// assumed that the encoding is simple: e.g. uses the same OID for
+	// signature and public key like Ed25519.
+	Oid() asn1.ObjectIdentifier
+}
+
+// Additional methods when the signature scheme is supported in TLS.
+type TLSScheme interface {
+	TLSIdentifier() uint
+}
+
+// SchemeByName returns the scheme with the given name and nil if it is not
+// supported.  Use ListSchemes() to list supported schemes.
+// Names are case insensitive.
+func SchemeByName(name string) Scheme {
+	// XXX add a (compile time?) lookup table
+	name = strings.ToLower(name)
+	for _, scheme := range schemes {
+		if strings.ToLower(scheme.Name()) == name {
+			return scheme
+		}
+	}
+	return nil
+}
+
+// ListSchemeNames returns the names of all schemes supported.
+func ListSchemeNames() []string {
+	ret := []string{}
+	for _, scheme := range schemes {
+		ret = append(ret, scheme.Name())
+	}
+	return ret
+}
+
+func SchemeByOid(oid asn1.ObjectIdentifier) Scheme {
+	// XXX add a (compile time?) lookup table
+	for _, scheme := range schemes {
+		certScheme, ok := scheme.(CertificateScheme)
+		if !ok {
+			continue
+		}
+		if certScheme.Oid().Equal(oid) {
+			return scheme
+		}
+	}
+	return nil
+}
+
+func SchemeByTLSIdentifier(id uint) Scheme {
+	// XXX add a (compile time?) lookup table
+	for _, scheme := range schemes {
+		tlsScheme, ok := scheme.(TLSScheme)
+		if !ok {
+			continue
+		}
+		if tlsScheme.TLSIdentifier() == id {
+			return scheme
+		}
+	}
+	return nil
+}
+
+func UnmarshalPEMPublicKey(data []byte) (PublicKey, error) {
+	block, rest := pem.Decode(data)
+	if len(rest) != 0 {
```

Comment:
I think the checks are for the block, like in the [Decode example](https://golang.org/pkg/encoding/pem/#Decode)

```
        block, rest := pem.Decode(pubPEMData)
	if block == nil || block.Type != "PUBLIC KEY" {
		log.Fatal("failed to decode PEM block containing public key")
	}
```